File: DevExpress-Demo/VB/NavigationDemo.Wpf/ViewModels/CustomizedContentViewModel.vb
Imports System
Imports System.Collections.Generic
Imports System.IO
Imports System.Linq
Imports System.Windows.Threading
Imports System.Xml.Serialization
Imports DevExpress.Internal
Imports DevExpress.Mvvm.DataAnnotations
Imports DevExpress.Mvvm.POCO
Imports DevExpress.Xpf.DemoBase.Helpers

Namespace NavigationDemo

    <POCOViewModel>
    Public Class CustomizedContentViewModel

        Const HistoryLength As Integer = 20

        Const UpdateInterval As Integer = 1000

        Private random As Random = New Random()

        Public Property Data As IEnumerable(Of StockItemInfo)

        Public Sub New()
            Dim data = StockItems.GetData().[Select](Function(x) StockItemInfo.Create(x.CompanyName, x.Price, x.Volume, x.LowPrice, x.HighPrice)).ToList()
            For Each stockItem In data
                For i As Integer = 0 To HistoryLength - 1
                    UpdatePrice(stockItem)
                Next
            Next

            Me.Data = data
        End Sub

        Private timer As DispatcherTimer

        Public Sub OnLoaded()
            timer = New DispatcherTimer() With {.Interval = TimeSpan.FromMilliseconds(UpdateInterval)}
            AddHandler timer.Tick, AddressOf timer_Tick
            timer.Start()
        End Sub

        Public Sub OnUnloaded()
            If timer IsNot Nothing Then
                timer.Stop()
                RemoveHandler timer.Tick, AddressOf timer_Tick
                timer = Nothing
            End If
        End Sub

        Private Sub timer_Tick(ByVal sender As Object, ByVal e As EventArgs)
            For Each stockItem In Data
                UpdatePrice(stockItem)
            Next
        End Sub

        Private Sub UpdatePrice(ByVal stockItem As StockItemInfo)
            Dim newDelta As Double = random.NextDouble() * 0.5 - 0.25
            If stockItem.Price + newDelta <= 0 Then newDelta = 0
            If Math.Sign(stockItem.DeltaPrice) = Math.Sign(newDelta) Then
                stockItem.DeltaChange = 0
            Else
                stockItem.DeltaChange = Math.Sign(newDelta)
            End If

            stockItem.DeltaPrice = newDelta
            stockItem.Price += stockItem.DeltaPrice
            If stockItem.Price < stockItem.LowPrice Then stockItem.LowPrice = stockItem.Price
            If stockItem.Price > stockItem.HighPrice Then stockItem.HighPrice = stockItem.Price
            stockItem.DeltaPricePercent = stockItem.DeltaPrice / stockItem.Price
            stockItem.Volume += Convert.ToInt32(random.NextDouble() * stockItem.Volume * 0.005 - 0.0025)
            UpdatePriceHistory(stockItem)
        End Sub

        Private Sub UpdatePriceHistory(ByVal stockItem As StockItemInfo)
            Dim newPriceHistory As List(Of Double) = New List(Of Double)(New Double(19) {})
            For i As Integer = 1 To HistoryLength - 1
                newPriceHistory(i - 1) = If(i < stockItem.PriceHistory.Count, stockItem.PriceHistory(i), 0)
            Next

            newPriceHistory(HistoryLength - 1) = stockItem.Price
            stockItem.PriceHistory = newPriceHistory
        End Sub
    End Class

    <XmlRoot("StockItems")>
    Public Class StockItems
        Inherits List(Of StockItem)

        Public Shared Function GetData() As List(Of StockItem)
            Dim s As XmlSerializer = New XmlSerializer(GetType(StockItems))
            Dim result As List(Of StockItem) = Nothing
            Dim path = DataDirectoryHelper.GetFile("StockSource.xml", DataDirectoryHelper.DataFolderName)
            Using fs As FileStream = File.OpenRead(path)
                result = CType(s.Deserialize(fs), List(Of StockItem))
            End Using

            Return result
        End Function
    End Class

    Public Class StockItem

        Public Property CompanyName As String

        Public Property Price As Double

        Public Property Volume As Integer

        Public Property LowPrice As Double

        Public Property HighPrice As Double
    End Class

    <POCOViewModel>
    Public Class StockItemInfo

        Public Shared Function Create(ByVal companyName As String, ByVal price As Double, ByVal volume As Integer, ByVal lowPrice As Double, ByVal highPrice As Double) As StockItemInfo
            Dim itemInfo = ViewModelSource.Create(Function() New StockItemInfo())
            itemInfo.CompanyName = companyName
            itemInfo.Price = price
            itemInfo.Volume = volume
            itemInfo.LowPrice = lowPrice
            itemInfo.HighPrice = highPrice
            Return itemInfo
        End Function

        Public Overridable Property CompanyName As String

        Public Overridable Property Price As Double

        Public Overridable Property Volume As Integer

        Public Overridable Property LowPrice As Double

        Public Overridable Property HighPrice As Double

        Public Overridable Property DeltaPrice As Double

        Public Overridable Property DeltaPricePercent As Double

        Public Overridable Property DeltaChange As Integer

        Protected Sub New()
            PriceHistory = New List(Of Double)()
        End Sub

        Public Overridable Property PriceHistory As List(Of Double)
    End Class
End Namespace

File: DevExpress-Demo/CS/NavigationDemo.Wpf/ViewModels/CustomizedContentViewModel.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Windows.Threading;
using System.Xml.Serialization;
using DevExpress.Internal;
using DevExpress.Mvvm.DataAnnotations;
using DevExpress.Mvvm.POCO;
using DevExpress.Xpf.DemoBase.Helpers;

namespace NavigationDemo {
    [POCOViewModel]
    public class CustomizedContentViewModel {
        const int HistoryLength = 20;
        const int UpdateInterval = 1000;
        Random random = new Random();

        public IEnumerable<StockItemInfo> Data { get; set; }

        public CustomizedContentViewModel() {
            var data = StockItems.GetData().Select(x => StockItemInfo.Create(x.CompanyName, x.Price, x.Volume, x.LowPrice, x.HighPrice)).ToList();
            foreach(var stockItem in data)
                for(int i = 0; i < HistoryLength; i++)
                    UpdatePrice(stockItem);
            Data = data;
        }

        DispatcherTimer timer;
        public void OnLoaded() {
            timer = new DispatcherTimer() { Interval = TimeSpan.FromMilliseconds(UpdateInterval) };
            timer.Tick += timer_Tick;
            timer.Start();
        }
        public void OnUnloaded() {
            if(timer != null) {
                timer.Stop();
                timer.Tick -= timer_Tick;
                timer = null;
            }
        }
        void timer_Tick(object sender, EventArgs e) {
            foreach(var stockItem in Data)
                UpdatePrice(stockItem);
        }

        void UpdatePrice(StockItemInfo stockItem) {
            double newDelta = random.NextDouble() * 0.5 - 0.25;
            if(stockItem.Price + newDelta <= 0)
                newDelta = 0;
            if(Math.Sign(stockItem.DeltaPrice) == Math.Sign(newDelta))
                stockItem.DeltaChange = 0;
            else
                stockItem.DeltaChange = Math.Sign(newDelta);
            stockItem.DeltaPrice = newDelta;
            stockItem.Price += stockItem.DeltaPrice;
            if(stockItem.Price < stockItem.LowPrice)
                stockItem.LowPrice = stockItem.Price;
            if(stockItem.Price > stockItem.HighPrice)
                stockItem.HighPrice = stockItem.Price;
            stockItem.DeltaPricePercent = stockItem.DeltaPrice / stockItem.Price;
            stockItem.Volume += Convert.ToInt32(random.NextDouble() * stockItem.Volume * 0.005 - 0.0025);
            UpdatePriceHistory(stockItem);
        }
        void UpdatePriceHistory(StockItemInfo stockItem) {
            List<double> newPriceHistory = new List<double>(new double[HistoryLength]);
            for(int i = 1; i < HistoryLength; i++)
                newPriceHistory[i - 1] = i < stockItem.PriceHistory.Count ? stockItem.PriceHistory[i] : 0;
            newPriceHistory[HistoryLength - 1] = stockItem.Price;
            stockItem.PriceHistory = newPriceHistory;
        }
    }

    [XmlRoot("StockItems")]
    public class StockItems : List<StockItem> {
        public static List<StockItem> GetData() {
            XmlSerializer s = new XmlSerializer(typeof(StockItems));
            List<StockItem> result = null;
            var path = DataDirectoryHelper.GetFile("StockSource.xml", DataDirectoryHelper.DataFolderName);
            using(FileStream fs = File.OpenRead(path))
                result = (List<StockItem>)s.Deserialize(fs);
            return result;
        }
    }

    public class StockItem {
        public string CompanyName { get; set; }
        public double Price { get; set; }
        public int Volume { get; set; }
        public double LowPrice { get; set; }
        public double HighPrice { get; set; }
    }
    [POCOViewModel]
    public class StockItemInfo {
        public static StockItemInfo Create(string companyName, double price, int volume, double lowPrice, double highPrice) {
            var itemInfo = ViewModelSource.Create(() => new StockItemInfo());
            itemInfo.CompanyName = companyName;
            itemInfo.Price = price;
            itemInfo.Volume = volume;
            itemInfo.LowPrice = lowPrice;
            itemInfo.HighPrice = highPrice;
            return itemInfo;
        }

        public virtual string CompanyName { get; set; }
        public virtual double Price { get; set; }
        public virtual int Volume { get; set; }
        public virtual double LowPrice { get; set; }
        public virtual double HighPrice { get; set; }
        public virtual double DeltaPrice { get; set; }
        public virtual double DeltaPricePercent { get; set; }
        public virtual int DeltaChange { get; set; }

        protected StockItemInfo() {
            PriceHistory = new List<double>();
        }
        public virtual List<double> PriceHistory { get; set; }
    }
}
